composer/src/main/kotlin/com/gojuno/composer/TestRun.kt

```diff
@@ -194,17 +194,14 @@ private fun saveLogcat(adbDevice: AdbDevice, logsDir: File): Observable<Pair<Str
                             false -> "${previous.logcat}\n$newline"
                         }
 
-                        fun String.parseTestClassAndName(prefix: String): Pair<String, String>? = this.substringAfter(prefix, missingDelimiterValue = "").let {
-                            when (it) {
-                                "" -> null
-                                else -> it.substringAfter("(").removeSuffix(")") to it.substringBefore("(")
-                            }
+                        fun String.parseTestClassAndName(regex: Regex): Pair<String, String>? {
+                            return regex.find(this)?.destructured?.let { (_, testMethod, testClass) -> Pair(testMethod, testClass) }
                         }
 
                         // Implicitly expecting to see logs from `android.support.test.internal.runner.listener.LogRunListener`.
                         // Was not able to find more reliable solution to capture logcat per test.
-                        val startedTest: Pair<String, String>? = newline.parseTestClassAndName("TestRunner: started: ")
-                        val finishedTest: Pair<String, String>? = newline.parseTestClassAndName("TestRunner: finished: ")
+                        val startedTest: Pair<String, String>? = newline.parseTestClassAndName("""(.*TestRunner.*: started: )(.*)\((.*)\)""".toRegex())
```

Comment:
I wondered about perf, but personally find it very readable. It's easy to see the capture groups and destructuring makes it clear exactly what were pulling out of the log line.

I can give a non-regex implementation a shot too.

Reply:
If it's not hard, please try, would love to compare!

Reply:
Yeah, I'd compare it with non-regex solution, too

Reply:
How about:

```kotlin
fun String.parseTestClassAndName(): Pair<String, String>? {
    val tokens = split(':')
    if (tokens.size != 3) return null

    if (tokens[0].trimStart('I', '/').startsWith("TestRunner")) {
        if (tokens[1] == " started" || tokens[1] == " finished") {
            return tokens[2].substringAfter("(").removeSuffix(")") to tokens[2].substringBefore("(").trim()
        }
    }
    return null
}
```

I'm not sure about the performance of `split()`. Since it's a single character I'm assuming it's O(n) where n is the length of the string.

Reply:
Maybe the first check should be `if (!trimStart('I', '/').startsWith("TestRunner"))` so we can avoid O(n) on every log.